tlog-mirror.md

```diff
@@ -0,0 +1,297 @@
+request by internally updating the pending checkpoint and responding with an
+empty response body. The mirror MUST retain the log's signature in the pending
+checkpoint.
+
+The mirror cosigner MUST NOT sign the checkpoint in this process. It MAY respond
+with witness cosignatures if the mirror operator wishes to additionally provide
+a separate witness service using its pending checkpoint. If so, this witness
+service MUST be a distinct cosigner from the mirror cosigner, with a distinct
+name. The mirror's signature is computed later, as described below.
+
+### add-entries
+
+The mirror implements a `add-entries` endpoint to upload entries for a supported
+log:
+
+    POST <submission prefix>/<encoded origin>/add-entries
+
+`encoded origin` is the log's origin, [percent-encoded][].
+
+#### Request Body
+
+The request body MUST have `Content-Type` of `application/octet-stream` and
+contain the following values, concatenated.
+
+* 8 bytes, encoding a big-endian, unsigned 64-bit integer: `upload_start`
+* 8 bytes, encoding a big-endian, unsigned 64-bit integer: `upload_end`
+* 2 bytes, encoding a big-endian, unsigned 16-bit integer: `ticket_size`
+* `ticket_size` bytes, containing an opaque `ticket` value, described below
+* A sequence of *entry packages*, described below
+
+`upload_end` MUST be equal to the pending checkpoint's tree size, or that of a
+previously valid pending checkpoint. `ticket` is an opaque value from the
+mirror, or the empty string, to help the mirror recover past pending
+checkpoints.
+
+`upload_start` MUST be less or equal to `upload_end`. It MUST also be less or
+equal to the mirror's next entry for the origin.
+
+The request body uploads the log entries whose indices are in
+`[upload_start, upload_end)`. Entries are grouped into bounded-size entry
+packages. Each package has a [subtree consistency proof][] that allows the
+mirror to verify and commit the entries in the package without buffering the
+entire request body.
+
+Each entry package is determined by a half-open interval `[start, end)` of log
+indices. The request MUST contain entry packages whose intervals' disjoint
+union, in order, is the interval `[upload_start, upload_end)`. The overall
+interval MUST be divided into packages at multiples of 256.
+
+More precisely, if `upload_start` is equal to `upload_end`, there are no
+packages. Otherwise, let `rounded_start` be `upload_start` rounded down to a
+multiple of 256, and let `rounded_end` be `upload_end` rounded up to a
+multiple of 256. The request MUST contain
+`num_packages = (rounded_end - rounded_start) / 256` packages. Entry
+package `i`, for `0 <= i < num_packages`, MUST be computed from the interval
+`[start, end)` where:
+
+    start = upload_start, rounded_start + i * 256
+    end = upload_end, rounded_start + (i + 1) * 256
+
+The package MUST contain the following values, concatenated.
+
+* The log entries in `[start, end)`, each with a big-endian, 16-bit length
```

Comment:
Maybe we should tweak the tlog-tiles phrasing to use 16-bit instead of uint16 for consistency's sake: https://github.com/C2SP/C2SP/blob/main/tlog-tiles.md?plain=1#L136.

Reply:
Yeah, probably just a carryover from the first version of these specs being written for a Golang library.

Reply:
Looks like lots of C2SP specs all say uintN. I've switched to that just because it's shorter and honestly everyone will know what that means. Meh. :-)